index.bs

```diff
@@ -424,58 +433,150 @@ When this method is invoked, the user agent MUST execute the following algorithm
+    Issue(w3c/webauthn#259): The rest of this algorithm assumes |rpId| is an
+    [=origin=], but the above step sometimes produces a string.
+
+1. Let |normalizedParameters| be a new [=list=] whose [=list/items=] are pairs of
+    ScopedCredentialType and a [=dictionary=] type (as returned by [=normalizing
+    an algorithm=]).
+1. [=list/For each=] |current| of {{cryptoParameters}}:
+    1. If <code>|current|.{{ScopedCredentialParameters/type}}</code> does not
+        contain a {{ScopedCredentialType}} supported by this implementation,
+        then [=continue=].
+    1. Let |normalizedAlgorithm| be the result of <a>normalizing an algorithm</a>
+        [[!WebCryptoAPI]], with |alg| set to
+        <code>|current|.{{algorithm}}</code> and |op| set to `"generateKey"`. If
+        an error occurs during this procedure, then [=continue=].
+    1. [=list/Append=] the pair of
+        <code>|current|.{{ScopedCredentialParameters/type}}</code> and
+        |normalizedAlgorithm| to |normalizedParameters|.
+
+1. If |normalizedParameters| is empty and {{cryptoParameters}} was not empty,
+    cancel the timer started in step 2, return [=a promise rejected with=] with
+    a {{DOMException}} whose name is "{{NotSupportedError}}", and terminate this
+    algorithm.
+
+1. Let |clientExtensions| be a new [=list=].
+1. If the {{ScopedCredentialOptions/extensions}} member of {{options}} is
+    <a>present</a>, then [=map/for each=] |extension| → |argument| of
+    <code>{{options}}.{{ScopedCredentialOptions/extensions}}</code>:
+    1. If |extension| is not supported by this client platform, then either:
+        * [=Continue=], or
+        * Let |result| be a CBOR ([[!RFC7049]]) encoding of |extension|.
+
+            Issue(w3c/webauthn#363): Define this encoding more precisely.
+    1. Otherwise, let |result| be the result of running |extension|'s [=client
+        processing=] algorithm on |argument|. If the algorithm returned an
+        error, [=continue=].
+
+        Issue(w3c/webauthn#363): Ensure all extensions define a client
+        processing algorithm.
+    1. [=list/Append=] |result| to |clientExtensions|.
+
+1. Let |clientData| be a new {{ClientData}} instance whose fields are:
+    : {{challenge}}
+    :: The [=base64url encoding=] of {{attestationChallenge}}
+    : {{origin}}
+    :: The [=unicode serialization of an origin|unicode serialization=] of |rpId|
+    : {{hashAlg}}
+    :: UA-chosen
+
+        Issue(w3c/webauthn#362): We need *some* constraints on the possible hash
+        algorithms, or else sites will fail on unusual UAs.
+    : {{tokenBinding}}
+    :: The [=Token Binding ID=] associated with |callerOrigin| (if any)
+
+        Issue(w3c/webauthn#360): Make sure this association was set up properly.
+    : {{ClientData/extensions}}
+    :: |clientExtensions|
+
+1. Let |clientDataJSON| be the [=UTF-8 encoding=] of the result of calling the
+    initial value of {{JSON/stringify|JSON.stringify}} on |clientData|.
```

Comment:
You might want to make this change centrally in https://w3c.github.io/webauthn/#clientdata-clientdatajson and reference that here. Maybe we should also change the variable name so that we can say something like "Generate the [=clientDataJSON=] and call it |J|."